File: tools/server/internal/archive/albumthumbs_test.go
package archive

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"server/internal/js"
	ts "server/internal/testsupport"
	"strings"
	"testing"
	"time"
)

// The tag appears in three shapes across the hosts the archive links to, and a
// relative URL must never be stored, so the extractor is compared against the
// JavaScript on every shape at once.
func TestExtractOgImageMatchesNode(t *testing.T) {
	pages := []string{
		`<meta property="og:image" content="https://live.staticflickr.com/1/2_b.jpg">`,
		`<meta content="https://lh3.googleusercontent.com/x" property="og:image">`,
		`<meta name="twitter:image" content="https://example.org/t.png">`,
		`<meta property='og:image' content='https://example.org/single.png'>`,
		`<META PROPERTY="OG:IMAGE" CONTENT="https://example.org/shouty.png">`,
		// Relative and non-http are refused rather than stored.
		`<meta property="og:image" content="/relative.png">`,
		`<meta property="og:image" content="data:image/png;base64,AAA">`,
		`<meta property="og:image" content="  https://example.org/padded.png  ">`,
		// The first shape wins when a page carries several.
		`<meta name="twitter:image" content="https://example.org/tw.png">` +
			`<meta property="og:image" content="https://example.org/og.png">`,
		`<html><head><title>no tag here</title></head>`,
		``,
	}
	payload := js.Arr()
	for _, page := range pages {
		payload.Append(js.Str(page))
	}
	out := ts.RunNode(t, nil, `
		const { extractOgImage } = await import('./scripts/lib/albumThumbs.js');
		const pages = JSON.parse(process.argv[1]);
		process.stdout.write(JSON.stringify(pages.map(extractOgImage)));
	`, string(payload.Encode("")))

	var want []string
	if err := json.Unmarshal([]byte(out), &want); err != nil {
		t.Fatal(err)
	}
	for i, page := range pages {
		if got := ExtractOgImage(page); got != want[i] {
			t.Errorf("ExtractOgImage(%q) = %q, node = %q", page, got, want[i])
		}
	}
}

// The rule that decides whether a cache is poisoned.
func TestDuplicateCoversMatchesNode(t *testing.T) {
	cache := js.Obj().
		Set("https://a", js.Obj().Set("thumb", js.Str("https://img/shared"))).
		Set("https://b", js.Obj().Set("thumb", js.Str("https://img/shared"))).
		Set("https://c", js.Obj().Set("thumb", js.Str("https://img/shared"))).
		Set("https://d", js.Obj().Set("thumb", js.Str("https://img/pair"))).
		Set("https://e", js.Obj().Set("thumb", js.Str("https://img/pair"))).
		Set("https://f", js.Obj().Set("thumb", js.Str("https://img/lonely"))).
		Set("https://g", js.Obj().Set("thumb", js.Str(""))).
		Set("https://h", js.Obj())
	albums := []string{"https://a", "https://b", "https://c", "https://d",
		"https://e", "https://f", "https://g", "https://h", "https://never-cached"}

	list := js.Arr()
	for _, url := range albums {
		list.Append(js.Str(url))
	}
	out := ts.RunNode(t, nil, `
		const { duplicateCovers } = await import('./scripts/lib/albumThumbs.js');
		const [cache, albums] = [JSON.parse(process.argv[1]), JSON.parse(process.argv[2])];
		process.stdout.write(JSON.stringify(
			Object.fromEntries(duplicateCovers(cache, albums))));
	`, string(cache.Encode("")), string(list.Encode("")))

	var want map[string][]string
	if err := json.Unmarshal([]byte(out), &want); err != nil {
		t.Fatal(err)
	}
	got := DuplicateCovers(cache, albums, 3)
	if len(got) != len(want) {
		t.Fatalf("go found %v, node found %v", got, want)
	}
	for thumb, urls := range want {
		if strings.Join(got[thumb], ",") != strings.Join(urls, ",") {
			t.Errorf("%s: go %v, node %v", thumb, got[thumb], urls)
		}
	}
	// The pair must NOT be flagged: two albums can honestly share a photo.
	if _, flagged := got["https://img/pair"]; flagged {
		t.Error("two albums sharing a cover is not evidence of a poisoned cache")
	}
}

// The allowlist is the only thing stopping this endpoint fetching a URL of the
// caller's choosing, so it is read from real dataset files.
func TestKnownAlbumsReadsDatasets(t *testing.T) {
	root := t.TempDir()
	events := filepath.Join(root, "events", "drupaljam", "nl")
	if err := os.MkdirAll(events, 0o755); err != nil {
		t.Fatal(err)
	}
	write := func(name, body string) {
		if err := os.WriteFile(filepath.Join(events, name), []byte(body), 0o644); err != nil {
			t.Fatal(err)
		}
	}
	write("2024.json", `{"event":{"flickr":{"groupUrl":"  https://flickr.com/groups/x  "}}}`)
	write("2023.json", `{"event":{"flickr":{"groupUrl":"https://photos.google.com/share/y"}}}`)
	write("2022.json", `{"event":{"name":"no album"}}`)
	write("broken.json", `{"event":`)
	write("notes.txt", `https://not-a-dataset.example`)

	albums, err := KnownAlbums(root)
	if err != nil {
		t.Fatal(err)
	}
	want := []string{"https://flickr.com/groups/x", "https://photos.google.com/share/y"}
	if got := SortedAlbums(albums); strings.Join(got, ",") != strings.Join(want, ",") {
		t.Errorf("albums = %v, want %v", got, want)
	}

	// A dataset that will not parse must not take the whole allowlist with it.
	if len(albums) != 2 {
		t.Errorf("a broken dataset changed the allowlist: %v", albums)
	}
}

// A missing directory is the normal state of a fresh checkout, not a failure.
func TestKnownAlbumsToleratesNoEvents(t *testing.T) {
	albums, err := KnownAlbums(t.TempDir())
	if err != nil || len(albums) != 0 {
		t.Errorf("= %v %v", albums, err)
	}
}

func TestThumbCacheRoundTripsThroughNode(t *testing.T) {
	dir := t.TempDir()
	if cache := LoadThumbCache(dir); cache.Len() != 0 {
		t.Error("no cache file yet should read as empty, not as an error")
	}

	cache := js.Obj().
		Set("https://flickr.com/groups/x", js.Obj().
			Set("thumb", js.Str("https://live.staticflickr.com/1/2_b.jpg")).
			Set("at", js.Int(1700000000000))).
		Set("https://photos.google.com/share/y", js.Obj().
			Set("thumb", js.Str("")).
			Set("at", js.Int(1700000001000)))
	if err := SaveThumbCache(dir, cache); err != nil {
		t.Fatal(err)
	}

	// Byte-identical to what the JavaScript writes for the same object.
	written, err := os.ReadFile(filepath.Join(dir, "album-thumbs.json"))
	if err != nil {
		t.Fatal(err)
	}
	nodeText := ts.RunNode(t, nil, `
		process.stdout.write(JSON.stringify(JSON.parse(process.argv[1]), null, 2) + '\n');
	`, string(cache.Encode("")))
	if string(written) != nodeText {
		t.Errorf("cache file differs:\ngo:   %q\nnode: %q", written, nodeText)
	}

	reloaded := LoadThumbCache(dir)
	if reloaded.Get("https://flickr.com/groups/x").Get("thumb").StrVal() !=
		"https://live.staticflickr.com/1/2_b.jpg" {
		t.Errorf("reloaded = %s", reloaded.Encode(""))
	}
}

// A cache file somebody has hand-edited into invalid JSON should cost a refetch,
// not a crash.
func TestThumbCacheSurvivesCorruption(t *testing.T) {
	dir := t.TempDir()
	if err := os.WriteFile(filepath.Join(dir, "album-thumbs.json"), []byte("{oops"), 0o644); err != nil {
		t.Fatal(err)
	}
	if cache := LoadThumbCache(dir); cache.Len() != 0 {
		t.Errorf("= %s", cache.Encode(""))
	}
}

func TestResolveThumbCachingRules(t *testing.T) {
	var hits int
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		hits++
		if agent := r.Header.Get("user-agent"); !strings.Contains(agent, "conference-planner") {
			t.Errorf("user-agent = %q", agent)
		}
		if r.URL.Path == "/missing" {
			w.WriteHeader(http.StatusTooManyRequests)
			return
		}
		// The tag sits behind a lot of markup, as it does on a real album page.
		_, _ = w.Write([]byte("<html><head>" + strings.Repeat("<link>", 5000) +
			`<meta property="og:image" content="https://img.example/cover.jpg">` +
			"</head><body>ignored</body></html>"))
	}))
	defer server.Close()

	cache := js.Obj()
	now := time.Unix(1700000000, 0)
	options := ThumbOptions{Client: server.Client(), Now: now}

	thumb, cached := ResolveThumb(server.URL+"/album", cache, options)
	if thumb != "https://img.example/cover.jpg" || cached {
		t.Fatalf("first = %q %v", thumb, cached)
	}
	if hits != 1 {
		t.Fatalf("hits = %d", hits)
	}

	// A hit is served from the cache for thirty days.
	options.Now = now.Add(29 * 24 * time.Hour)
	if thumb, cached = ResolveThumb(server.URL+"/album", cache, options); !cached || hits != 1 {
		t.Errorf("day 29: %q cached=%v hits=%d", thumb, cached, hits)
	}
	options.Now = now.Add(31 * 24 * time.Hour)
	if _, cached = ResolveThumb(server.URL+"/album", cache, options); cached || hits != 2 {
		t.Errorf("day 31 should re-ask: cached=%v hits=%d", cached, hits)
	}

	// A miss is cached too, but for two days rather than thirty — a 429 is
	// usually the host's mood, not the album's fault.
	options.Now = now
	if thumb, cached = ResolveThumb(server.URL+"/missing", cache, options); thumb != "" || cached {
		t.Errorf("miss = %q %v", thumb, cached)
	}
	before := hits
	options.Now = now.Add(24 * time.Hour)
	if _, cached = ResolveThumb(server.URL+"/missing", cache, options); !cached || hits != before {
		t.Errorf("day 1 should still be cached: cached=%v", cached)
	}
	options.Now = now.Add(3 * 24 * time.Hour)
	if _, cached = ResolveThumb(server.URL+"/missing", cache, options); cached || hits != before+1 {
		t.Errorf("day 3 should re-ask a miss: cached=%v hits=%d", cached, hits)
	}

	// Refresh is how a rate-limited miss gets cleared without waiting.
	options.Now = now
	before = hits
	if _, cached = ResolveThumb(server.URL+"/album", cache, options); !cached {
		t.Fatal("expected a cached hit to set up the refresh case")
	}
	options.Refresh = true
	if _, cached = ResolveThumb(server.URL+"/album", cache, options); cached || hits != before+1 {
		t.Errorf("refresh should re-ask: cached=%v hits=%d", cached, hits)
	}
}

// An unreachable host is an answer, not an error.
func TestResolveThumbSurvivesADeadHost(t *testing.T) {
	cache := js.Obj()
	thumb, cached := ResolveThumb("http://127.0.0.1:1/album", cache,
		ThumbOptions{Now: time.Unix(1700000000, 0)})
	if thumb != "" || cached {
		t.Errorf("= %q %v", thumb, cached)
	}
	// The failure is recorded so the next page view does not retry immediately.
	if !cache.Get("http://127.0.0.1:1/album").Exists() {
		t.Error("a failed resolution should still be cached as a miss")
	}
}

// The ceiling exists so a stranger's endless document cannot be read forever.
func TestReadOgImageGivesUpAtTheCeiling(t *testing.T) {
	tag := `<meta property="og:image" content="https://img.example/deep.jpg">`
	page := "<html><head>" + strings.Repeat("<link>", 30000) + tag

	if got := readOgImage(strings.NewReader(page), 2_000_000); got != "https://img.example/deep.jpg" {
		t.Errorf("within the ceiling: %q", got)
	}
	// Cut the budget below where the tag sits and it is simply not found.
	if got := readOgImage(strings.NewReader(page), 1000); got != "" {
		t.Errorf("past the ceiling: %q", got)
	}
}

File: scripts/lib/__tests__/albumThumbs.test.js
// Resolving an album's cover means fetching a page from someone else's server,
// so the two things worth pinning are what we read out of it and what we refuse
// to fetch at all.
import { describe, it, expect, beforeAll, afterAll, vi } from 'vitest';
import { mkdtemp, mkdir, writeFile, rm } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { extractOgImage, knownAlbums, resolveThumb, duplicateCovers } from '../albumThumbs.js';

describe('extractOgImage', () => {
  it('reads og:image in either attribute order', () => {
    expect(extractOgImage('<meta property="og:image" content="https://x.test/a.jpg">')).toBe(
      'https://x.test/a.jpg',
    );
    expect(extractOgImage("<meta content='https://x.test/b.jpg' property='og:image'>")).toBe(
      'https://x.test/b.jpg',
    );
  });

  it('falls back to twitter:image when there is no og:image', () => {
    expect(extractOgImage('<meta name="twitter:image" content="https://x.test/c.jpg">')).toBe(
      'https://x.test/c.jpg',
    );
  });

  it('ignores a relative image, which we could not resolve later', () => {
    expect(extractOgImage('<meta property="og:image" content="/local/a.jpg">')).toBe('');
  });

  it('says nothing rather than guessing', () => {
    expect(extractOgImage('<html><body>no meta here</body></html>')).toBe('');
    expect(extractOgImage('')).toBe('');
    expect(extractOgImage(null)).toBe('');
  });
});

describe('knownAlbums', () => {
  let dir;
  beforeAll(async () => {
    dir = await mkdtemp(join(tmpdir(), 'albums-'));
    await mkdir(join(dir, 'events', 'nested'), { recursive: true });
    await writeFile(
      join(dir, 'events', 'a.json'),
      JSON.stringify({ event: { flickr: { groupUrl: 'https://flic.kr/g/abc' } }, items: [] }),
    );
    await writeFile(
      join(dir, 'events', 'nested', 'b.json'),
      JSON.stringify({
        event: { flickr: { groupUrl: 'https://photos.app.goo.gl/xyz' } },
        items: [],
      }),
    );
    // No album, and a file that will not parse: neither may contribute a URL.
    await writeFile(join(dir, 'events', 'c.json'), JSON.stringify({ event: {}, items: [] }));
    await writeFile(join(dir, 'events', 'broken.json'), '{ not json');
  });
  afterAll(async () => {
    if (dir) await rm(dir, { recursive: true, force: true });
  });

  it('collects every recorded album, including nested datasets', async () => {
    const set = await knownAlbums(dir);
    expect([...set].sort()).toEqual(['https://flic.kr/g/abc', 'https://photos.app.goo.gl/xyz']);
  });

  it('is the allowlist: anything not in the archive is simply absent', async () => {
    const set = await knownAlbums(dir);
    // This is what stops the endpoint being pointed at an arbitrary host.
    expect(set.has('http://169.254.169.254/latest/meta-data/')).toBe(false);
    expect(set.has('https://flickr.com/groups/not-ours/')).toBe(false);
  });
});

describe('resolveThumb', () => {
  const page =
    '<html><head><meta property="og:image" content="https://live.staticflickr.com/1.jpg"></head></html>';

  it('fetches once and serves the cache afterwards', async () => {
    const cache = {};
    const fetchImpl = vi.fn(async () => ({ ok: true, text: async () => page }));
    const first = await resolveThumb('https://flic.kr/g/abc', { cache, fetchImpl });
    expect(first).toEqual({ thumb: 'https://live.staticflickr.com/1.jpg', cached: false });

    const second = await resolveThumb('https://flic.kr/g/abc', { cache, fetchImpl });
    expect(second).toEqual({ thumb: 'https://live.staticflickr.com/1.jpg', cached: true });
    expect(fetchImpl).toHaveBeenCalledTimes(1);
  });

  it('remembers a miss, so a coverless album is not re-fetched every time', async () => {
    const cache = {};
    const fetchImpl = vi.fn(async () => ({ ok: true, text: async () => '<html></html>' }));
    expect((await resolveThumb('https://x.test/none', { cache, fetchImpl })).thumb).toBe('');
    expect((await resolveThumb('https://x.test/none', { cache, fetchImpl })).cached).toBe(true);
    expect(fetchImpl).toHaveBeenCalledTimes(1);
  });

  it('gives up on a miss far sooner than on a hit', async () => {
    // Resolving 33 albums in a row had Google rate-limit two of them; both had
    // covers seconds later. A miss must not be believed for a month.
    const fetchImpl = vi.fn(async () => ({ ok: true, text: async () => page }));
    const missCache = { 'https://x.test/a': { thumb: '', at: 0 } };
    expect(
      (await resolveThumb('https://x.test/a', { cache: missCache, fetchImpl, now: 3 * 864e5 }))
        .cached,
    ).toBe(false);

    const hitCache = { 'https://x.test/b': { thumb: 'https://old.test/x.jpg', at: 0 } };
    expect(
      (await resolveThumb('https://x.test/b', { cache: hitCache, fetchImpl, now: 3 * 864e5 }))
        .cached,
    ).toBe(true);
  });

  it('refetches once the cached answer is old enough to be worth rechecking', async () => {
    const cache = { 'https://flic.kr/g/abc': { thumb: 'https://old.test/x.jpg', at: 0 } };
    const fetchImpl = vi.fn(async () => ({ ok: true, text: async () => page }));
    const r = await resolveThumb('https://flic.kr/g/abc', { cache, fetchImpl, now: 40 * 864e5 });
    expect(r.cached).toBe(false);
    expect(r.thumb).toBe('https://live.staticflickr.com/1.jpg');
  });

  it('treats an unreachable host as "no cover", not as a failure', async () => {
    const cache = {};
    const fetchImpl = vi.fn(async () => {
      throw new Error('ETIMEDOUT');
    });
    await expect(resolveThumb('https://x.test/down', { cache, fetchImpl })).resolves.toEqual({
      thumb: '',
      cached: false,
    });
  });
});

// A cache holding the WRONG cover is worse than one holding none: a hit lives for
// thirty days, so a bad write is served for a month and filling gaps never
// touches it. Production showed one photograph on all 54 cards until somebody
// noticed by eye — this is the check that makes it self-correcting.
describe('duplicateCovers', () => {
  const at = 1;
  const cache = (pairs) => Object.fromEntries(pairs.map(([u, t]) => [u, { thumb: t, at }]));

  it('flags a cover claimed by three or more albums', () => {
    const same = 'https://live.staticflickr.com/same.jpg';
    const urls = ['a', 'b', 'c', 'd'];
    const dupes = duplicateCovers(
      cache([
        ['a', same],
        ['b', same],
        ['c', same],
        ['d', 'https://x/1.jpg'],
      ]),
      urls,
    );
    expect([...dupes.keys()]).toEqual([same]);
    expect(dupes.get(same)).toEqual(['a', 'b', 'c']);
  });

  it('leaves two albums sharing a cover alone — the same photo can be in two pools', () => {
    const same = 'https://live.staticflickr.com/same.jpg';
    expect(
      duplicateCovers(
        cache([
          ['a', same],
          ['b', same],
        ]),
        ['a', 'b'],
      ).size,
    ).toBe(0);
  });

  it('ignores albums with no cover, however many there are', () => {
    const c = cache([
      ['a', ''],
      ['b', ''],
      ['c', ''],
      ['d', ''],
    ]);
    expect(duplicateCovers(c, ['a', 'b', 'c', 'd']).size).toBe(0);
  });

  it('does not flag a healthy cache', () => {
    const c = cache([
      ['a', 'https://x/1.jpg'],
      ['b', 'https://x/2.jpg'],
      ['c', 'https://x/3.jpg'],
    ]);
    expect(duplicateCovers(c, ['a', 'b', 'c']).size).toBe(0);
  });
});
